rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,374 @@
+
+### Out of scope
+
+Out of scope are:
+
+- improvements to the build time of Vrl (mainly caused by lalrpop).
+- changes to improve the functionality of Vrl.
+
+## Pain
+
+### Allocations
+
+A lot of performance is taken up by excessive allocations. It is likely that
+reducing the number of allocations will produce the biggest performance gain
+for VRL and thus this should be the main area of focus.
+
+Where we have clones that we should try to avoid:
+
+- At the start of the Remap transform. The entire event is cloned so we can
+  roll back to it on error.
+- In the interface between VRL and Vector. The object is cloned on both `get`
+  and `set` operations.
+
+Let's have a look at the life of the event `{ "foo": { "bar": 42 } }` in this
+simple VRL source:
+
+```coffee
+  .zub = .foo
+```
+
+1. The whole event is cloned in order to create a backup in case the event is
+   aborted. [here](https://github.com/vectordotdev/vector/blob/v0.17.0/src/transforms/remap.rs#L135)
+2. Then we need to access `.foo`. The `.foo` field needs to be pulled out of
+   the event and converted to a `vrl::Value`. [here](https://github.com/vectordotdev/vector/blob/v0.17.0/lib/vector-core/src/event/vrl_target.rs#L130)
+   This actually involves two allocations.
+   a. First the value is cloned.
+   b. Next it is converted into a `vrl::Value`.
+      [here](https://github.com/vectordotdev/vector/blob/v0.17.0/lib/vector-core/src/event/value.rs#L308)
+      Since `.foo` in at Object, this involves allocating a new `BTreeMap` then
+      looping over each field, allocating a new `vrl::Value` for each one.
+3. We then assign it to `.zub`. To do this the value is cloned and inserted
+   into our event. [here](https://github.com/vectordotdev/vector/blob/v0.17.0/lib/vrl/compiler/src/expression/assignment.rs#L369)
+
+Each time an object is passed between Vector and VRL an allocation is made.
+
+### Boxing
+
+VRL works by compiling the source to a tree based AST (abstract syntax tree).
+To execute the program the tree is walked, each node is evaluated with the
+current state of the program. Each node is not necessarily stored in memory
+close to the next node. As a result it does not fully utilise the cpu cache.
+
+Also, each node in the AST is boxed which causes extra pointer indirection.
+
+
+## Proposal
+
+### Implementation
+
+### Reducing allocations
+
+#### Use reference counting
+
+One of the data types that is cloned the most is `vrl::Value`. A lot of these
+clones are because VRL needs to maintain multiple references to a single value.
+In the Rust model, this can get highly awkward handling the lifetimes.
+
+If VRL used reference counting instead, the allocations can be avoided. Each
+clone would just increase a count into the underlying data store which is
+significantly cheaper. When the value goes out of scope the count is reduced.
+If the count reaches 0, the memory is freed.
+
+VRL needs to wrap the data in an `Rc`. This does mean pulling the data out of
```

Comment:
What would the advantages of a `Cow` be? I can think mutation would be safer since it copies rather than panics. Anything else?

@bruceg is on the mark, ownership gets complicated. Ultimately the event store should own all the data, but the data is not always created there, it is often created in an expression and then added to the event store - sometimes whilst also passing it somewhere else.

Reply:
Looks like there's no advantages but this might be worthwhile spelling out in an alternatives section.